### PR TITLE
Feature/build free text query for advanced search criteria

### PR DESCRIPTION
**Pull request by PenghaiZhang**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [ ] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

This PR adds the back-end re-implementation of how to support Advanced search criteria for New Search UI.

Key changes are:

1. Create data structure on server to provide information required to build a `FreeTextQuery` for each Wizard control. 
2. Add a variation of `search2` endpoint to support a POST request  in which the request body provides data in above structure.
3. Create functions to build a `FreeTextQuery` based on the provided data.
4. Combine the `FreeTextQuery` with other search criteria in `ItemIndex.java`.
5. Add new test cases for a variety of Wizard control combinations.

### CODE REVIEW

File: autotest/OldTests/src/test/java/io/github/openequella/rest/Search2ApiAdditionalTest.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+import org.apache.commons.httpclient.methods.StringRequestEntity;
+import org.codehaus.jackson.JsonNode;
+import org.codehaus.jackson.node.ArrayNode;
+import org.codehaus.jackson.node.ObjectNode;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+/** This test focuses on the variation of search2 which supports Advanced search criteria. */
+@TestInstitution("fiveo")
+public class Search2ApiAdditionalTest extends AbstractRestApiTest {
```
Comment (edalex-ian) [inline autotest/OldTests/src/test/java/io/github/openequella/rest/Search2ApiAdditionalTest.java:24]: It looks as though there's no testing of having both the body and also some normal search params - like even just limiting to a collection and/or status.

**Review by edalex-ian — CHANGES_REQUESTED** [review]
Wow, great turnaround there @PenghaiZhang :+1: Very quick! :runner: 

I think we need to shuffle a few things - one is naming, but the bigger one is whether we can avoid putting things into DefaultSearch (or if we have to, making it more generic).

Have a read, happy to discuss as needed.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/model/AdvancedSearchCriteria.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+case class WizardControlFieldValue(schemaNodes: Array[String],
+                                   values: Array[String],
+                                   queryType: String)
+
+/**
+  * Provides a list of `WizardControlFieldValue` to help build the full criteria.
+  *
+  * @param controlValues A list of `WizardControlFieldValue`.
+  */
+case class AdvancedSearchCriteria(controlValues: Array[WizardControlFieldValue])
```
[the remark this answers is not in the document]
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/model/AdvancedSearchCriteria.scala:44]: Good idea! I have update the names.

File: Source/Plugins/Core/com.equella.base/src/com/tle/common/search/DefaultSearch.java [hunk 3]
```diff
@@ -70,20 +70,20 @@
   protected Collection<DateFilter> dateFilters = Lists.newArrayList();
   private final List<List<Field>> musts = new ArrayList<List<Field>>();
   private final List<List<Field>> mustNots = new ArrayList<List<Field>>();
-  private String customLuceneQuery;
+  private FreeTextQuery advancedSearchCriteria;
```
[the remark this answers is not in the document]
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.base/src/com/tle/common/search/DefaultSearch.java:73]: I have deleted this field.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala [hunk 4]
```diff
@@ -120,7 +127,7 @@ object SearchHelper {
       case None    => whereQuery
     }
     search.setFreeTextQuery(freeTextQuery)
-    search.setCustomLuceneQuery(params.customLuceneQuery)
+    search.setAdvancedSearchCriteria(fieldValues.map(buildAdvancedSearchCriteria).orNull)
```
[the remark this answers is not in the document]
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala:130]: Yes, I have achieved this by some new code.

File: autotest/OldTests/src/test/java/io/github/openequella/rest/Search2ApiAdditionalTest.java (hunk 1 above)
Comment (PenghaiZhang) [reply autotest/OldTests/src/test/java/io/github/openequella/rest/Search2ApiAdditionalTest.java:24]: There is one test case `searchWithOtherParamsTest` :smile:

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala [hunk 5]
```diff
@@ -115,19 +115,24 @@ object SearchHelper {
...
-      case None    => whereQuery
-    }
+    val dynaCollectionQuery: Option[FreeTextBooleanQuery] = handleDynaCollection(
+      params.dynaCollection)
+    val whereQuery: Option[FreeTextBooleanQuery] = Option(params.whereClause).map(WhereParser.parse)
+    val advSearchCriteria: Option[FreeTextBooleanQuery] =
+      fieldValues.map(buildAdvancedSearchCriteria)
+
+    val freeTextQuery: FreeTextBooleanQuery =
+      List(dynaCollectionQuery, whereQuery, advSearchCriteria)
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala:125]: Looks like another approach is `flatten`.

Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala:125]: Oh but still need to do a `concat`.

Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala:125]: Forgot to add a comment for this.

```
    // Merge above three optional FreeTextBooleanQuery into one. If both queries are defined, combine them.
    // If only one is non-empty, return it. If both are empty, return None.
```

Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala:125]: I am trying `cats - semigroup` now.

**Review by edalex-ian — APPROVED** [review]
Love ya work @PenghaiZhang . Very cool to see how building our FP foundations with fp-ts is now paying off also over in Scala. Hopefully though we don't forget that `|+|` is from cats.

Also interesting to see your use of `implicit` there. Nice. :+1: 

When you get a green build, merge on.

**Pull request merged**